Enoch/ac 65 post product transaction

ExtraGravity opened the pull request:
APIs have their own interface types, database has their own, service in the middle to translate between the two types.

Example CURL request
```JSON
curl -d '{
	"datetime": "2000-01-01 12:12:12Z",
	"toPersonUuid": "a2b121fd-1a1f-4425-97db-876af3c5bd2f",
	"fromPersonUuid": "2fc0977d-c49b-47e8-9a03-090c5c65792d",
	"amountOfProduct": 99.99,
	"costPerUnit": 99.99,
	"currency": "UGX",
	"milkQuality": "9999"
}' -H "Content-Type: application/json" -X POST http://localhost:9090/transactions/products/milk

```

ExtraGravity committed:
omg post requset into db, thank the lord

ExtraGravity committed:
remove unncessary changes

ExtraGravity committed:
ac-65 abstract out database and API interfaces

Use separate interfaces for database and API types.
Service is used to translate between the two.
Implemented for GET product transactions, post has been commented out.

ExtraGravity committed:
Merge branch 'enoch/ac-65-post-product-transaction' of github.com:agriculture-capstone/agricore into enoch/ac-65-post-product-transaction

ExtraGravity committed:
ac-66 fix post product transaction with new strategy

Posting product transactions use completely separate interfaces for the
database and the API. A service is in the middle to translate between
the two formats.

beaesguerra commented on src/routers/transactions/products.ts:
```diff
@@ -89,8 +121,24 @@ router.get('/:type', async (req, res) => {
+  }
+
+  const createResult = await ProdTransactionsService.createProdTransactionsInDb(createReq);
+  const uuid = createResult[0];
```
how come createResult is an array? might be better to return the first element instead of the whole array in createProdTransactionsInDb so that it cannot be misused

ExtraGravity committed:
ac-65 fix linter errors, mostly add documentation

ExtraGravity committed:
ac-65 handle some invalid requests

Only handles missing requests right now. Uncaught errors will send a 500
status code.

beaesguerra commented:
who generates the datetime? the client or the server?

beaesguerra commented on src/services/ProductTransactions.ts:
```diff
@@ -55,14 +88,34 @@ export async function createProdTransactionsInDb(apiReq: api.ProdTransactionReq)
+  try {
+    newUuid = await db.insertProdTransaction(dbInsertReq, attributes);
+  } catch (e) {
+    throw new Error(unhandledErrorMsg);
```
should we also be including the original error message from 'e' so that it is easier for us to debug?

ExtraGravity replied:
I'll log it, but it shouldn't be thrown up, because the clients can see it.

beaesguerra commented on src/services/ProductTransactions.ts:
```diff
@@ -40,12 +43,42 @@ export async function getProdTransactionsFromDb(productType: string): Promise<ap
+  let productTypeId: number = -1;
+  const invalidFields: string[] = [];
+
+  // validate request
```
no big deal but maybe it'll be worth while putting this in a function? so then the update function can use it too to validate the fields

ExtraGravity committed:
ac-65 add uuid to required parameter in posting a new product transaction.

The agritrader could create a farmer offline, and link a new product transaction
to that new farmer which is not in the database. If the database sends a new uuid then
the agritrader must update all the product transactions. Requiring a uuid and using that
one makes the agritrader alot easier.

ExtraGravity commented:
Replaced by #74
